gms/gossiper.cc

```diff
@@ -182,8 +182,10 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
         co_return;
     }
 
-    /* If the message is from a node with a different group0 id throw it away. */
-    if (syn_msg.group0_id() && get_group0_id() && syn_msg.group0_id() != get_group0_id()) {
+    // If the message is from a node with a different group0 id throw it away.
+    // A group0 id mismatch is expected during a rolling restart in the Raft-based recovery procedure.
+    if (_gcfg.recovery_leader().empty()
+            && syn_msg.group0_id() && get_group0_id() && syn_msg.group0_id() != get_group0_id()) {
         logger.warn("Group0Id mismatch from {} {} != {}", from.addr, syn_msg.group0_id(), get_group0_id());
```

Comment:
If we see a group0 mismatch, perhaps we should log it (as info, with a different message) as the first sign that recovery is beginning.

Reply:
We could, but this message would be continually logged during the whole rolling restart. It would flood the log files.

We could also log it only once or once per contact point, but that would require storing some state in gossiper just for a single log.

We already have logs that are the first sign that recovery is beginning, logged during restart in `join_cluster` after setting `recovery_leader` (see the first patch). I think these are enough.

Reply:
Perhaps add recovery_leader to the syn message as an optional? Then we can look into it for a match, if group0 id doesn't match.